source/opt/local_single_store_elim_pass.cpp

```diff
@@ -0,0 +1,445 @@
+      ip->opcode() == SpvOpStore ?  kSpvStorePtrId : kSpvLoadPtrId);
+  ir::Instruction* ptrInst = def_use_mgr_->GetDef(*varId);
+  ir::Instruction* varInst = ptrInst;
+  while (IsNonPtrAccessChain(varInst->opcode())) {
+    *varId = varInst->GetSingleWordInOperand(kSpvAccessChainPtrId);
+    varInst = def_use_mgr_->GetDef(*varId);
+  }
+  return ptrInst;
+}
+
+bool LocalSingleStoreElimPass::IsTargetVar(uint32_t varId) {
+  if (seen_non_target_vars_.find(varId) != seen_non_target_vars_.end())
+    return false;
+  if (seen_target_vars_.find(varId) != seen_target_vars_.end())
+    return true;
+  const ir::Instruction* varInst = def_use_mgr_->GetDef(varId);
+  assert(varInst->opcode() == SpvOpVariable);
+  const uint32_t varTypeId = varInst->type_id();
+  const ir::Instruction* varTypeInst = def_use_mgr_->GetDef(varTypeId);
+  if (varTypeInst->GetSingleWordInOperand(kSpvTypePointerStorageClass) !=
+    SpvStorageClassFunction) {
+    seen_non_target_vars_.insert(varId);
+    return false;
+  }
+  const uint32_t varPteTypeId =
+    varTypeInst->GetSingleWordInOperand(kSpvTypePointerTypeId);
+  ir::Instruction* varPteTypeInst = def_use_mgr_->GetDef(varPteTypeId);
+  if (!IsTargetType(varPteTypeInst)) {
+    seen_non_target_vars_.insert(varId);
+    return false;
+  }
+  seen_target_vars_.insert(varId);
+  return true;
+}
+
+void LocalSingleStoreElimPass::SingleStoreAnalyze(ir::Function* func) {
+  ssa_var2store_.clear();
+  non_ssa_vars_.clear();
+  store2idx_.clear();
+  store2blk_.clear();
+  for (auto bi = func->begin(); bi != func->end(); ++bi) {
+    uint32_t instIdx = 0;
+    for (auto ii = bi->begin(); ii != bi->end(); ++ii, ++instIdx) {
+      switch (ii->opcode()) {
+      case SpvOpStore: {
+        // Verify store variable is target type
+        uint32_t varId;
+        ir::Instruction* ptrInst = GetPtr(&*ii, &varId);
+        if (non_ssa_vars_.find(varId) != non_ssa_vars_.end())
+          continue;
+        if (IsNonPtrAccessChain(ptrInst->opcode())) {
+          non_ssa_vars_.insert(varId);
+          ssa_var2store_.erase(varId);
+          continue;
+        }
+        // Verify target type and function storage class
+        if (!IsTargetVar(varId)) {
+          non_ssa_vars_.insert(varId);
+          continue;
+        }
+        // If already stored, disqualify it
+        if (ssa_var2store_.find(varId) != ssa_var2store_.end()) {
+          non_ssa_vars_.insert(varId);
+          ssa_var2store_.erase(varId);
+          continue;
+        }
+        // Remember iterator of variable's store and it's
+        // ordinal position in function
+        ssa_var2store_[varId] = &*ii;
+        store2idx_[&*ii] = instIdx;
+        store2blk_[&*ii] = &*bi;
+      } break;
+      case SpvOpFunctionCall: {
+        // For now, empty SSA variable set and terminate analysis
+        // TODO(): Add logic to more optimally handle function calls.
+        ssa_var2store_.clear();
+        return;
+      } break;
+      default:
```

Comment:
I appreciate you catching these.

For the moment, my preference is to completely bail on seeing OpAtomics, Modf, Frexp or OpCopyMemory and leave better handling of these as a TODO.

My reading of OpLifetime* is that these are optimization hints and can be safely ignored. Let me know if you disagree.

Re OpSelect, according to the spec validation rules, OpSelect is not a designated consumer or producer of pointers under logical addressing. This pass bails on non-logical addressing modules.

I will address OpCopyObject in the following comment.